### ISSUE TITLE
Report transitively unused variables for `no-unused-vars`.

### ISSUE THREAD

**Issue by futpib** [eslint/eslint#10742]
**What rule do you want to change?**

[no-unused-vars](https://eslint.org/docs/rules/no-unused-vars)

**Does this change cause the rule to produce more or fewer warnings?**

**More** (but if you consider multiple executions of eslint in process of fixing unused vars in a project, you could say it stays the **same**)

**How will the change be implemented? (New option, new default behavior, etc.)?**

Probably a new option (might call it `transitive`).

**Please provide some example code that this change will affect:**

```js
// example 1
const a = 0;
const b = a + 1;
```

```js
// example 2
const a = () => b;
const b = () => a;
```

**What does the rule currently do for this code?**

(example 1) It reports `b` as an unused variable.
(example 2) It reports nothing.

**What will the rule do after it's changed?**

For both examples 1 and 2 It would report both `a` and `b` as unused variables, since they are used only in initializing expressions of other (transitively) unused variable declarations.

**Comment by kaicataldo** [comment]
Hi, sorry for the delay in responding to this.

> It would report both a and b as unused variables as long as they are used only in initializing expressions of other (transitively) unused variable declarations.

Mind updating this so that it's clear which example this is referring to?

**Comment by futpib** [comment]
I meat to refer to both examples, I edited the issue accordingly.
